gateway-js/src/supergraphManagers/UplinkSupergraphManager/index.ts

```diff
@@ -0,0 +1,294 @@
+    this.logger = logger ?? getDefaultLogger(debug);
+
+    this.uplinkEndpoints = uplinkEndpoints ?? this.uplinkEndpoints;
+    this.maxRetries = maxRetries ?? this.uplinkEndpoints.length * 3 - 1;
+
+    this.fetcher = fetcher ?? this.fetcher;
+    this.fallbackPollIntervalMs =
+      fallbackPollIntervalInMs ?? this.fallbackPollIntervalMs;
+    this.shouldRunSubgraphHealthcheck =
+      shouldRunSubgraphHealthcheck ?? this.shouldRunSubgraphHealthcheck;
+    this.onFailureToFetchSupergraphSdlDuringInit =
+      onFailureToFetchSupergraphSdlDuringInit;
+    this.onFailureToFetchSupergraphSdlAfterInit =
+      onFailureToFetchSupergraphSdlAfterInit;
+
+    this.state = { phase: 'constructed' };
+  }
+
+  public async initialize({ update, healthCheck }: SupergraphSdlHookOptions) {
+    this.update = update;
+
+    if (this.shouldRunSubgraphHealthcheck) {
+      this.healthCheck = healthCheck;
+    }
+
+    let initialSupergraphSdl: string | undefined = undefined;
+    try {
+      const result = await this.updateSupergraphSdl();
+      if (!result) {
+        throw new Error(
+          'Invalid supergraph schema supplied during initialization.',
+        );
+      }
+      initialSupergraphSdl = result.supergraphSdl;
+      if (result?.minDelaySeconds) {
+        this.minDelayMs = 1000 * result?.minDelaySeconds;
+        this.earliestFetchTime = new Date(Date.now() + this.minDelayMs);
+      }
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    this.state = { phase: 'initialized' };
+
+    // Start polling after we resolve the first supergraph
+    this.beginPolling();
+
+    return {
+      supergraphSdl: initialSupergraphSdl,
+      cleanup: async () => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          clearTimeout(this.timerRef);
+          this.timerRef = null;
+        }
+      },
+    };
+  }
+
+  public async nextFetch(): Promise<void | null> {
+    if (this.state.phase !== 'polling') {
+      return;
+    }
+    return this.state.nextFetchPromise;
+  }
+
+  private async updateSupergraphSdl(): Promise<{
+    supergraphSdl: string;
+    minDelaySeconds?: number;
+  } | null> {
+    let supergraphSdl;
+    let minDelaySeconds: number | undefined =
+      this.fallbackPollIntervalMs / 1000;
+
+    try {
+      const result = await loadSupergraphSdlFromUplinks({
+        graphRef: this.graphRef,
+        apiKey: this.apiKey,
+        endpoints: this.uplinkEndpoints,
+        errorReportingEndpoint: this.errorReportingEndpoint,
+        fetcher: this.fetcher,
+        compositionId: this.compositionId ?? null,
+        maxRetries: this.maxRetries,
+        roundRobinSeed: this.fetchCount++,
+        earliestFetchTime: this.earliestFetchTime,
+        logger: this.logger,
+      });
+
+      if (!result) {
+        return null;
+      }
+
+      this.compositionId = result.id;
```

Comment:
What is this used for?

Reply:
It's passed into subsequent requests so we only get updates after this `compositionId` https://github.com/apollographql/federation/blob/cc7b6dabc54a261d7071d9331f06d1250cda8357/gateway-js/src/supergraphManagers/UplinkSupergraphManager/loadSupergraphSdlFromStorage.ts#L122-L129

Reply:
Lemme know if there's anything else unclear around this